web/source/kmwosk.js

```diff
@@ -1500,25 +1670,93 @@ if(!window['tavultesoft']['keymanweb']['initialized']) {
      * @return                                Always true
      * Description  Update the current shift state within KMW
      */
-    osk._UpdateVKShift = function(e, v, d)
-    {
-      var keyShiftState=0;
-      if(e)
+    osk._UpdateVKShift = function(e, v, d) {
+      var keyShiftState=0, lockStates=0, i;
+
+      var lockNames  = ["CAPS", "NUM_LOCK", "SCROLL_LOCK"];
+      var lockKeys   = ["K_CAPS", "K_NUMLOCK", "K_SCROLL"];
+
+      if(e) {
         // read shift states from Pevent
-        keyShiftState = e.Lmodifiers/0x10;
-      else if(d)
+        keyShiftState = e.Lmodifiers;
+        lockStates = e.Lstates;
+
+        for(i=0; i < lockNames.length; i++) {
+          if(lockStates & osk.stateBitmasks[lockNames[i]]) {
+            osk._stateKeys[lockKeys[i]] = lockStates & osk.modifierCodes[lockNames[i]];
+          }
+        }
+      } else if(d) {
         keyShiftState |= v;
-      else
+
+        for(i=0; i < lockNames.length; i++) {
+          if(v & osk.stateBitmasks[lockNames[i]]) {
+            osk._stateKeys[lockKeys[i]] = true;
+          }
+        }
+      } else {
         keyShiftState &= ~v;
 
+        for(i=0; i < lockNames.length; i++) {
+          if(v & osk.stateBitmasks[lockNames[i]]) {
+            osk._stateKeys[lockKeys[i]] = false;
+          }
+        }
+      }
+
       // Find and display the selected OSK layer
       osk.layerId=osk.getLayerId(keyShiftState);
-      if(osk._Visible) osk._Show();
+
+      // osk._UpdateVKShiftStyle will be called automatically upon the next _Show.
+      if(osk._Visible) {
+        osk._Show();
+      }
 
       return true;
     }
 
     /**
+     * Function     _UpdateVKShiftStyle
+     * Scope        Private
+     * @param       {string=}   layerId
+     * Description  Updates the OSK's visual style for any toggled state keys
+     */
+    osk._UpdateVKShiftStyle = function(layerId) {
+      var i, n, layer=null, layerElement=null;
+
+      if(layerId) {
+        for(n=0; n<osk.layers.length; n++) {
+          if(osk.layers[n]["id"] == osk.layerId) {
+            break;
+          }
+        }
+
+        return;  // Failed to find the requested layer.
+      } else {
+        n=osk.layerIndex;
+        layerId=osk.layers[n]["id"];
+      }
+
+      layer=osk.layers[n];
+      
+      // Set the on/off state of any visible state keys.
+      var states =["K_CAPS",      "K_NUMLOCK",  "K_SCROLL"];
+      var keys   =[layer.capsKey, layer.numKey, layer.scrollKey];
+
+      for(i=0; i < keys.length; i++) {
+        // Skip any keys not in the OSK!
+        if(keys[i] == null) {
+          continue;
+        }
+
+        keys[i]['sp'] = osk._stateKeys[states[i]] ? '2' : '1';
```

Comment:
Man I'd love to use constants for the `'sp'` values

Reply:
That could be arranged.  There's a clear definition array in `osk.setButtonClass` that can be extracted into a set of constants.

Reply:
Also, numbers-in-strings ftw but let's not fix that today.